The job **`dist-x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3975562119/jobs/6815498957) [(plain)](https://github.com/rust-lang/rust/commit/2355a179c4615d017272329f476d4b5dd378d850/checks/10797702669/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  98% (38648/39314)
Updating files:  99% (38921/39314)
Updating files: 100% (39314/39314)
Updating files: 100% (39314/39314), done.
Note: switching to 'refs/remotes/pull/103019/merge'.
You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

---
  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at c4871b08 Merge 2355a179c4615d017272329f476d4b5dd378d850 into 21f683935257713eae8549e8b328367006097053
[command]"C:\Program Files\Git\bin\git.exe" log -1 --format='%H'
'c4871b084d99d4c8eedb0bf0c00ba671f27aa53b'
'c4871b084d99d4c8eedb0bf0c00ba671f27aa53b'
##[group]Run echo "[CI_PR_NUMBER=$num]"
echo "[CI_PR_NUMBER=$num]"
env:
  CI_JOB_NAME: dist-x86_64-msvc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  EXTRA_VARIABLES: {
 "RUST_CONFIGURE_ARGS": "--build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin",
 "SCRIPT": "PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths",
}
##[endgroup]
adding extra environment variable DIST_REQUIRE_ALL_TOOLS
adding extra environment variable RUST_CONFIGURE_ARGS
---
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  DIST_REQUIRE_ALL_TOOLS: 1
  RUST_CONFIGURE_ARGS: --build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin
  SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
##[endgroup]
Executing the job since there is no skip rule preventing the execution
##[group]Run src/ci/scripts/verify-channel.sh
src/ci/scripts/verify-channel.sh
---
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  DIST_REQUIRE_ALL_TOOLS: 1
  RUST_CONFIGURE_ARGS: --build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin
  SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
##[endgroup]
##[group]Run rust-lang/simpleinfra/github-actions/cancel-outdated-builds@master
with:
  github_token: ***
---
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  DIST_REQUIRE_ALL_TOOLS: 1
  RUST_CONFIGURE_ARGS: --build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin
  SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
##[endgroup]
Successfully daemonized cancel-outdated-builds.
##[group]Run src/ci/scripts/collect-cpu-stats.sh
src/ci/scripts/collect-cpu-stats.sh
---
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  DIST_REQUIRE_ALL_TOOLS: 1
  RUST_CONFIGURE_ARGS: --build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin
  SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
##[endgroup]
---
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  DIST_REQUIRE_ALL_TOOLS: 1
  RUST_CONFIGURE_ARGS: --build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin
  SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
##[endgroup]
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

---
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  DIST_REQUIRE_ALL_TOOLS: 1
  RUST_CONFIGURE_ARGS: --build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin
  SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
##[endgroup]
##[group]Run src/ci/scripts/install-clang.sh
src/ci/scripts/install-clang.sh
shell: C:\Program Files\Git\bin\bash.EXE --noprofile --norc -e -o pipefail {0}
shell: C:\Program Files\Git\bin\bash.EXE --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: dist-x86_64-msvc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  DIST_REQUIRE_ALL_TOOLS: 1
  RUST_CONFIGURE_ARGS: --build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin
  SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
##[endgroup]
Attempting with retry: curl -f https://ci-mirrors.rust-lang.org/rustc/LLVM-14.0.5-win64.exe -o LLVM-14.0.5-win64.exe
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
---
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  DIST_REQUIRE_ALL_TOOLS: 1
  RUST_CONFIGURE_ARGS: --build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin --set llvm.clang-cl=/d/a/rust/rust/citools/clang-rust/bin/clang-cl.exe
  SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
  NO_DOWNLOAD_CI_LLVM: 1
##[endgroup]
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
---
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  DIST_REQUIRE_ALL_TOOLS: 1
  RUST_CONFIGURE_ARGS: --build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin --set llvm.clang-cl=/d/a/rust/rust/citools/clang-rust/bin/clang-cl.exe
  SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
  NO_DOWNLOAD_CI_LLVM: 1
  WIX: /d/a/rust/rust/wix
##[endgroup]
---
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  DIST_REQUIRE_ALL_TOOLS: 1
  RUST_CONFIGURE_ARGS: --build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin --set llvm.clang-cl=/d/a/rust/rust/citools/clang-rust/bin/clang-cl.exe
  SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
  NO_DOWNLOAD_CI_LLVM: 1
  WIX: /d/a/rust/rust/wix
##[endgroup]
##[group]Run src/ci/scripts/install-mingw.sh
---
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  DIST_REQUIRE_ALL_TOOLS: 1
  RUST_CONFIGURE_ARGS: --build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin --set llvm.clang-cl=/d/a/rust/rust/citools/clang-rust/bin/clang-cl.exe
  SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
  NO_DOWNLOAD_CI_LLVM: 1
  WIX: /d/a/rust/rust/wix
##[endgroup]
warning: mingw-w64-x86_64-binutils-2.39-2 is up to date -- skipping
---
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  DIST_REQUIRE_ALL_TOOLS: 1
  RUST_CONFIGURE_ARGS: --build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin --set llvm.clang-cl=/d/a/rust/rust/citools/clang-rust/bin/clang-cl.exe
  SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
  NO_DOWNLOAD_CI_LLVM: 1
  WIX: /d/a/rust/rust/wix
##[endgroup]
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
---
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  DIST_REQUIRE_ALL_TOOLS: 1
  RUST_CONFIGURE_ARGS: --build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin --set llvm.clang-cl=/d/a/rust/rust/citools/clang-rust/bin/clang-cl.exe --enable-ninja
  SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
  NO_DOWNLOAD_CI_LLVM: 1
  WIX: /d/a/rust/rust/wix
##[endgroup]
##[group]Run src/ci/scripts/disable-git-crlf-conversion.sh
---
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  DIST_REQUIRE_ALL_TOOLS: 1
  RUST_CONFIGURE_ARGS: --build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin --set llvm.clang-cl=/d/a/rust/rust/citools/clang-rust/bin/clang-cl.exe --enable-ninja
  SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
  NO_DOWNLOAD_CI_LLVM: 1
  WIX: /d/a/rust/rust/wix
##[endgroup]
##[group]Run src/ci/scripts/verify-line-endings.sh
---
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  DIST_REQUIRE_ALL_TOOLS: 1
  RUST_CONFIGURE_ARGS: --build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin --set llvm.clang-cl=/d/a/rust/rust/citools/clang-rust/bin/clang-cl.exe --enable-ninja
  SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
  NO_DOWNLOAD_CI_LLVM: 1
  WIX: /d/a/rust/rust/wix
##[endgroup]
file:C:/Program Files/Git/etc/gitconfig diff.astextplain.textconv=astextplain
---
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  DIST_REQUIRE_ALL_TOOLS: 1
  RUST_CONFIGURE_ARGS: --build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin --set llvm.clang-cl=/d/a/rust/rust/citools/clang-rust/bin/clang-cl.exe --enable-ninja
  SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
  NO_DOWNLOAD_CI_LLVM: 1
  WIX: /d/a/rust/rust/wix
##[endgroup]
Skipping. This is only run when merging to the beta or stable branches.
---
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  DIST_REQUIRE_ALL_TOOLS: 1
  RUST_CONFIGURE_ARGS: --build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin --set llvm.clang-cl=/d/a/rust/rust/citools/clang-rust/bin/clang-cl.exe --enable-ninja
  SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
  NO_DOWNLOAD_CI_LLVM: 1
  WIX: /d/a/rust/rust/wix
##[endgroup]
This script only works on the stable channel. Skipping the check.
---
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  DIST_REQUIRE_ALL_TOOLS: 1
  RUST_CONFIGURE_ARGS: --build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin --set llvm.clang-cl=/d/a/rust/rust/citools/clang-rust/bin/clang-cl.exe --enable-ninja
  SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
  NO_DOWNLOAD_CI_LLVM: 1
  WIX: /d/a/rust/rust/wix
  AWS_ACCESS_KEY_ID: 
  AWS_SECRET_ACCESS_KEY: 
---
LowFree:        25537640 kB
SwapTotal:       4718592 kB
SwapFree:        4718592 kB
+ PGO_HOST=x86_64-pc-windows-msvc
+ python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
stage-build INFO: Running multi-stage build using Python 3.11.1 (tags/v3.11.1:a7a450f, Dec  6 2022, 19:58:39) [MSC v.1934 64 bit (AMD64)]
stage-build INFO: Environment values
{ 'ALLUSERSPROFILE': 'C:\\ProgramData',
  'ANDROID_HOME': 'C:\\Android\\android-sdk',
  'ANDROID_NDK': 'C:\\Android\\android-sdk\\ndk\\25.1.8937393',
  'ANDROID_NDK_HOME': 'C:\\Android\\android-sdk\\ndk\\25.1.8937393',
  'ANDROID_NDK_LATEST_HOME': 'C:\\Android\\android-sdk\\ndk\\25.1.8937393',
  'ANDROID_NDK_ROOT': 'C:\\Android\\android-sdk\\ndk\\25.1.8937393',
  'ANDROID_SDK_ROOT': 'C:\\Android\\android-sdk',
  'ANT_HOME': 'C:\\ProgramData\\chocolatey\\lib\\ant\\tools\\apache-ant-1.10.13',
  'APPDATA': 'C:\\Users\\runneradmin\\AppData\\Roaming',
  'AWS_ACCESS_KEY_ID': '',
  'AWS_SECRET_ACCESS_KEY': '',
  'AZURE_EXTENSION_DIR': 'C:\\Program Files\\Common '
                         'Files\\AzureCliExtensionDirectory',
  'CABAL_DIR': 'C:\\cabal',
  'CACHE_DOMAIN': 'ci-caches.rust-lang.org',
  'CHOCOLATEYINSTALL': 'C:\\ProgramData\\chocolatey',
  'CHROMEWEBDRIVER': 'C:\\SeleniumWebDrivers\\ChromeDriver',
  'CI': 'true',
  'CI_JOB_NAME': 'dist-x86_64-msvc',
  'COBERTURA_HOME': 'C:\\cobertura-2.1.1',
  'COMMONPROGRAMFILES': 'C:\\Program Files\\Common Files',
  'COMMONPROGRAMFILES(X86)': 'C:\\Program Files (x86)\\Common Files',
  'COMMONPROGRAMW6432': 'C:\\Program Files\\Common Files',
  'COMPILETEST_NEEDS_ALL_LLVM_COMPONENTS': '1',
  'COMPUTERNAME': 'fv-az212-630',
  'COMSPEC': 'C:\\Windows\\system32\\cmd.exe',
  'CONDA': 'C:\\Miniconda',
  'DEPLOY': '1',
  'DEPLOYMENT_BASEPATH': 'C:\\actions',
  'DIST_REQUIRE_ALL_TOOLS': '1',
  'DOTNET_MULTILEVEL_LOOKUP': '0',
  'DRIVERDATA': 'C:\\Windows\\System32\\Drivers\\DriverData',
  'EDGEWEBDRIVER': 'C:\\SeleniumWebDrivers\\EdgeDriver',
  'GCM_INTERACTIVE': 'Never',
  'GECKOWEBDRIVER': 'C:\\SeleniumWebDrivers\\GeckoDriver',
  'GHCUP_INSTALL_BASE_PREFIX': 'C:\\',
  'GHCUP_MSYS2': 'C:\\msys64',
  'GITHUB_ACTION': '__run_23',
  'GITHUB_ACTIONS': 'true',
  'GITHUB_ACTION_REF': '',
  'GITHUB_ACTION_REPOSITORY': '',
  'GITHUB_ACTOR': 'Kobzol',
  'GITHUB_ACTOR_ID': '4539057',
  'GITHUB_API_URL': 'https://api.github.com',
  'GITHUB_BASE_REF': 'master',
  'GITHUB_ENV': 'D:\\a\\_temp\\_runner_file_commands\\set_env_57d1ccd2-6b41-456b-a60f-bfa96c65e5e6',
  'GITHUB_EVENT_NAME': 'pull_request',
  'GITHUB_EVENT_PATH': 'D:\\a\\_temp\\_github_workflow\\event.json',
  'GITHUB_GRAPHQL_URL': 'https://api.github.com/graphql',
  'GITHUB_HEAD_REF': 'ci-multistage-python',
  'GITHUB_JOB': 'pr',
  'GITHUB_OUTPUT': 'D:\\a\\_temp\\_runner_file_commands\\set_output_57d1ccd2-6b41-456b-a60f-bfa96c65e5e6',
  'GITHUB_PATH': 'D:\\a\\_temp\\_runner_file_commands\\add_path_57d1ccd2-6b41-456b-a60f-bfa96c65e5e6',
  'GITHUB_REF': 'refs/pull/103019/merge',
  'GITHUB_REF_NAME': '103019/merge',
  'GITHUB_REF_PROTECTED': 'false',
  'GITHUB_REF_TYPE': 'branch',
  'GITHUB_REPOSITORY': 'rust-lang/rust',
  'GITHUB_REPOSITORY_ID': '724712',
  'GITHUB_REPOSITORY_OWNER': 'rust-lang',
  'GITHUB_REPOSITORY_OWNER_ID': '5430905',
  'GITHUB_RETENTION_DAYS': '90',
  'GITHUB_RUN_ATTEMPT': '1',
  'GITHUB_RUN_ID': '3975562119',
  'GITHUB_RUN_NUMBER': '81003',
  'GITHUB_SERVER_URL': 'https://github.com',
  'GITHUB_SHA': 'c4871b084d99d4c8eedb0bf0c00ba671f27aa53b',
  'GITHUB_STATE': 'D:\\a\\_temp\\_runner_file_commands\\save_state_57d1ccd2-6b41-456b-a60f-bfa96c65e5e6',
  'GITHUB_STEP_SUMMARY': 'D:\\a\\_temp\\_runner_file_commands\\step_summary_57d1ccd2-6b41-456b-a60f-bfa96c65e5e6',
  'GITHUB_TRIGGERING_ACTOR': 'Kobzol',
  'GITHUB_WORKFLOW': 'CI',
  'GITHUB_WORKFLOW_REF': 'rust-lang/rust/.github/workflows/ci.yml@refs/pull/103019/merge',
  'GITHUB_WORKFLOW_SHA': 'c4871b084d99d4c8eedb0bf0c00ba671f27aa53b',
  'GITHUB_WORKSPACE': 'D:\\a\\rust\\rust',
  'GOROOT_1_17_X64': 'C:\\hostedtoolcache\\windows\\go\\1.17.13\\x64',
  'GOROOT_1_18_X64': 'C:\\hostedtoolcache\\windows\\go\\1.18.10\\x64',
  'GOROOT_1_19_X64': 'C:\\hostedtoolcache\\windows\\go\\1.19.5\\x64',
  'GRADLE_HOME': 'C:\\ProgramData\\chocolatey\\lib\\gradle\\tools\\gradle-7.6',
  'HOME': 'C:\\msys64\\home\\runneradmin',
  'HOMEDRIVE': 'C:',
  'HOMEPATH': '\\Users\\runneradmin',
  'IEWEBDRIVER': 'C:\\SeleniumWebDrivers\\IEDriver',
  'IMAGEOS': 'win19',
  'IMAGEVERSION': '20230117.1',
  'JAVA_HOME': 'C:\\hostedtoolcache\\windows\\Java_Temurin-Hotspot_jdk\\8.0.352-8\\x64',
  'JAVA_HOME_11_X64': 'C:\\hostedtoolcache\\windows\\Java_Temurin-Hotspot_jdk\\11.0.17-8\\x64',
  'JAVA_HOME_13_X64': 'C:\\hostedtoolcache\\windows\\Java_Adopt_jdk\\13.0.2-8.1\\x64',
  'JAVA_HOME_17_X64': 'C:\\hostedtoolcache\\windows\\Java_Temurin-Hotspot_jdk\\17.0.5-8\\x64',
  'JAVA_HOME_8_X64': 'C:\\hostedtoolcache\\windows\\Java_Temurin-Hotspot_jdk\\8.0.352-8\\x64',
  'LOCALAPPDATA': 'C:\\Users\\runneradmin\\AppData\\Local',
  'LOGONSERVER': '\\\\fv-az212-630',
  'M2': 'C:\\ProgramData\\chocolatey\\lib\\maven\\apache-maven-3.8.7\\bin',
  'M2_REPO': 'C:\\ProgramData\\m2',
  'MAVEN_OPTS': '-Xms256m',
  'MIRRORS_BASE': 'https://ci-mirrors.rust-lang.org/rustc',
  'NO_DOWNLOAD_CI_LLVM': '1',
  'NPM_CONFIG_PREFIX': 'C:\\npm\\prefix',
  'NUMBER_OF_PROCESSORS': '8',
  'OS': 'Windows_NT',
  'PATH': 'D:\\a\\rust\\rust\\ninja;D:\\a\\rust\\rust\\msys2\\mingw64\\bin;C:\\hostedtoolcache\\windows\\Python\\3.11.1\\x64\\Scripts;C:\\hostedtoolcache\\windows\\Python\\3.11.1\\x64;C:\\msys64\\usr\\bin;D:\\a\\rust\\rust\\sccache;C:\\Program '
          'Files\\MongoDB\\Server\\5.0\\bin;C:\\aliyun-cli;C:\\vcpkg;C:\\cf-cli;C:\\Program '
          'Files (x86)\\NSIS;C:\\tools\\zstd;C:\\Program '
          'Files\\Mercurial;C:\\hostedtoolcache\\windows\\stack\\2.9.3\\x64;C:\\cabal\\bin;C:\\ghcup\\bin;C:\\tools\\ghc-9.4.4\\bin;C:\\Program '
          'Files\\dotnet;C:\\mysql\\bin;C:\\Program '
          'Files\\R\\R-4.2.2\\bin\\x64;C:\\SeleniumWebDrivers\\GeckoDriver;C:\\Program '
          'Files (x86)\\sbt\\bin;C:\\Program Files (x86)\\GitHub '
          'CLI;C:\\Program Files\\Git\\bin;C:\\Program Files '
          '(x86)\\pipx_bin;C:\\npm\\prefix;C:\\hostedtoolcache\\windows\\go\\1.17.13\\x64\\bin;C:\\hostedtoolcache\\windows\\Python\\3.7.9\\x64\\Scripts;C:\\hostedtoolcache\\windows\\Python\\3.7.9\\x64;C:\\hostedtoolcache\\windows\\Ruby\\2.5.9\\x64\\bin;C:\\tools\\kotlinc\\bin;C:\\hostedtoolcache\\windows\\Java_Temurin-Hotspot_jdk\\8.0.352-8\\x64\\bin;C:\\Program '
          'Files\\ImageMagick-7.1.0-Q16-HDRI;C:\\Program Files '
          '(x86)\\Microsoft '
          'SDKs\\Azure\\CLI2\\wbin;C:\\ProgramData\\kind;C:\\Program '
          'Files\\Eclipse '
          'Foundation\\jdk-8.0.302.8-hotspot\\bin;C:\\Windows\\system32;C:\\Windows;C:\\Windows\\System32\\Wbem;C:\\Windows\\System32\\WindowsPowerShell\\v1.0;C:\\Windows\\System32\\OpenSSH;C:\\ProgramData\\Chocolatey\\bin;C:\\Program '
          'Files\\PowerShell\\7;C:\\Program Files\\Microsoft\\Web Platform '
          'Installer;C:\\Program Files\\Microsoft SQL '
          'Server\\130\\Tools\\Binn;C:\\Program Files\\Microsoft SQL '
          'Server\\Client SDK\\ODBC\\170\\Tools\\Binn;C:\\Program Files '
          '(x86)\\Windows Kits\\10\\Windows Performance Toolkit;C:\\Program '
          'Files (x86)\\Microsoft SQL Server\\110\\DTS\\Binn;C:\\Program Files '
          '(x86)\\Microsoft SQL Server\\120\\DTS\\Binn;C:\\Program Files '
          '(x86)\\Microsoft SQL Server\\130\\DTS\\Binn;C:\\Program Files '
          '(x86)\\Microsoft SQL Server\\140\\DTS\\Binn;C:\\Program Files '
          '(x86)\\Microsoft SQL Server\\150\\DTS\\Binn;C:\\Program Files '
          '(x86)\\Microsoft SQL Server\\160\\DTS\\Binn;C:\\Program '
          'Files\\OpenSSL\\bin;C:\\Strawberry\\c\\bin;C:\\Strawberry\\perl\\site\\bin;C:\\Strawberry\\perl\\bin;C:\\ProgramData\\chocolatey\\lib\\pulumi\\tools\\Pulumi\\bin;C:\\Program '
          'Files\\TortoiseSVN\\bin;C:\\Program '
          'Files\\CMake\\bin;C:\\ProgramData\\chocolatey\\lib\\maven\\apache-maven-3.8.7\\bin;C:\\Program '
          'Files\\Microsoft Service '
          'Fabric\\bin\\Fabric\\Fabric.Code;C:\\Program Files\\Microsoft '
          'SDKs\\Service '
          'Fabric\\Tools\\ServiceFabricLocalClusterManager;C:\\Program '
          'Files\\nodejs;C:\\Program Files\\Git\\cmd;C:\\Program '
          'Files\\Git\\mingw64\\bin;C:\\Program '
          'Files\\Git\\usr\\bin;C:\\Program Files\\GitHub '
          'CLI;C:\\tools\\php;C:\\Program Files '
          '(x86)\\sbt\\bin;C:\\SeleniumWebDrivers\\ChromeDriver;C:\\SeleniumWebDrivers\\EdgeDriver;C:\\Program '
          'Files\\Amazon\\AWSCLIV2;C:\\Program '
          'Files\\Amazon\\SessionManagerPlugin\\bin;C:\\Program '
          'Files\\Amazon\\AWSSAMCLI\\bin;C:\\Program Files '
          '(x86)\\Google\\Cloud SDK\\google-cloud-sdk\\bin;C:\\Program Files '
          '(x86)\\Microsoft BizTalk Server;C:\\Program '
          'Files\\LLVM\\bin;C:\\Users\\runneradmin\\.dotnet\\tools;C:\\Users\\runneradmin\\.cargo\\bin;C:\\Users\\runneradmin\\AppData\\Local\\Microsoft\\WindowsApps',
  'PATHEXT': '.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC',
  'PERFLOG_LOCATION_SETTING': 'RUNNER_PERFLOG',
  'PGBIN': 'C:\\Program Files\\PostgreSQL\\14\\bin',
  'PGDATA': 'C:\\Program Files\\PostgreSQL\\14\\data',
  'PGO_HOST': 'x86_64-pc-windows-msvc',
  'PGPASSWORD': 'root',
  'PGROOT': 'C:\\Program Files\\PostgreSQL\\14',
  'PGUSER': 'postgres',
  'PHPROOT': 'c:\\tools\\php',
  'PIPX_BIN_DIR': 'C:\\Program Files (x86)\\pipx_bin',
  'PIPX_HOME': 'C:\\Program Files (x86)\\pipx',
  'POWERSHELL_DISTRIBUTION_CHANNEL': 'GitHub-Actions-win19',
  'POWERSHELL_UPDATECHECK': 'Off',
  'PROCESSOR_ARCHITECTURE': 'AMD64',
  'PROCESSOR_IDENTIFIER': 'Intel64 Family 6 Model 79 Stepping 1, GenuineIntel',
  'PROCESSOR_LEVEL': '6',
  'PROCESSOR_REVISION': '4f01',
  'PROGRAMDATA': 'C:\\ProgramData',
  'PROGRAMFILES': 'C:\\Program Files',
  'PROGRAMFILES(X86)': 'C:\\Program Files (x86)',
  'PROGRAMW6432': 'C:\\Program Files',
  'PSMODULEPATH': 'C:\\\\Modules\\azurerm_2.1.0;C:\\\\Modules\\azure_2.1.0;C:\\Users\\packer\\Documents\\WindowsPowerShell\\Modules;C:\\Program '
                  'Files\\WindowsPowerShell\\Modules;C:\\Windows\\system32\\WindowsPowerShell\\v1.0\\Modules;C:\\Program '
                  'Files\\Microsoft SQL '
                  'Server\\130\\Tools\\PowerShell\\Modules\\;C:\\Program Files '
                  '(x86)\\Google\\Cloud '
                  'SDK\\google-cloud-sdk\\platform\\PowerShell',
  'PUBLIC': 'C:\\Users\\Public',
  'PWD': 'D:/a/rust/rust',
  'RTOOLS40_HOME': 'C:\\rtools40',
  'RUNNER_ARCH': 'X64',
  'RUNNER_NAME': 'GitHub Actions 4',
  'RUNNER_OS': 'Windows',
  'RUNNER_PERFLOG': 'C:\\actions\\perflog',
  'RUNNER_TEMP': 'D:\\a\\_temp',
  'RUNNER_TOOL_CACHE': 'C:\\hostedtoolcache\\windows',
  'RUNNER_TRACKING_ID': 'github_59857113-f818-4cd7-a5d5-4408f7747f29',
  'RUNNER_WORKSPACE': 'D:\\a\\rust',
  'RUST_CONFIGURE_ARGS': '--build=x86_64-pc-windows-msvc '
                         '--host=x86_64-pc-windows-msvc '
                         '--target=x86_64-pc-windows-msvc --enable-full-tools '
                         '--enable-profiler --set rust.lto=thin --set '
                         'llvm.clang-cl=/d/a/rust/rust/citools/clang-rust/bin/clang-cl.exe '
                         '--enable-ninja --enable-sccache '
                         '--disable-manage-submodules --enable-locked-deps '
                         '--enable-cargo-native-static --set '
                         'rust.codegen-units-std=1 '
                         '--dist-compression-formats=xz --disable-dist-src '
                         '--release-channel=nightly '
                         '--enable-llvm-static-stdcpp --set '
                         'rust.remap-debuginfo --debuginfo-level-std=1 '
                         '--enable-missing-tools',
  'RUST_RELEASE_CHANNEL': 'nightly',
  'SBT_HOME': 'C:\\Program Files (x86)\\sbt\\',
  'SCCACHE_BUCKET': 'rust-lang-ci-sccache2',
  'SCRIPT': 'PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py '
            'python x.py dist bootstrap --include-default-paths',
  'SELENIUM_JAR_PATH': 'C:\\selenium\\selenium-server.jar',
  'SHLVL': '3',
  'SRC': '.',
  'STATS_RDCL': 'true',
  'SYSTEMDRIVE': 'C:',
  'SYSTEMROOT': 'C:\\Windows',
  'TEMP': 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp',
  'TERM': 'xterm-256color',
  'TMP': 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp',
  'TOOLSTATE_REPO': 'https://github.com/rust-lang-nursery/rust-toolstate',
  'TOOLSTATE_REPO_ACCESS_TOKEN': '',
  'USERDOMAIN': 'fv-az212-630',
  'USERDOMAIN_ROAMINGPROFILE': 'fv-az212-630',
  'USERNAME': 'runneradmin',
  'USERPROFILE': 'C:\\Users\\runneradmin',
  'VCPKG_INSTALLATION_ROOT': 'C:\\vcpkg',
  'VS140COMNTOOLS': 'C:\\Program Files (x86)\\Microsoft Visual Studio '
                    '14.0\\Common7\\Tools\\',
  'WINDIR': 'C:\\Windows',
  'WIX': 'D:/a/rust/rust/wix',
  '_': 'C:/hostedtoolcache/windows/Python/3.11.1/x64/python'}
stage-build INFO: Attempting to perform action `Download rustc-perf` with retry
stage-build INFO: Attempt 1/5
stage-build INFO: Downloading `https://github.com/rust-lang/rustc-perf/archive/3c253134664fdcba862c539d37f0de18557a9a4c.zip` into `\tmp\tmp-multistage\perf.zip`
stage-build DEBUG: Changing working dir from `D:\a\rust\rust` to `\tmp\tmp-multistage`
stage-build INFO: Unpacking archive `\tmp\tmp-multistage\perf.zip`
stage-build INFO: Moving `rustc-perf-3c253134664fdcba862c539d37f0de18557a9a4c` to `\tmp\tmp-multistage\rustc-perf`
stage-build INFO: Deleting file `\tmp\tmp-multistage\perf.zip`
stage-build DEBUG: Reverting working dir to `D:\a\rust\rust`
stage-build INFO: Stage `Build rustc (LLVM PGO)` starts
stage-build INFO: Executing `LLVM_PROFILE_DIR=\tmp\tmp-multistage\llvm-pgo\prof-%p C:\hostedtoolcache\windows\Python\3.11.1\x64\python.exe D:\a\rust\rust\x.py build --target x86_64-pc-windows-msvc --host x86_64-pc-windows-msvc --stage 2 library/std --llvm-profile-generate`
    Finished dev [unoptimized] target(s) in 0.16s
Building stage0 library artifacts (x86_64-pc-windows-msvc -> x86_64-pc-windows-msvc)
---
CMAKE_x86_64-pc-windows-msvc = None
CMAKE_x86_64_pc_windows_msvc = None
HOST_CMAKE = None
CMAKE = None
running: "cmake" "D:\\a\\rust\\rust\\src/llvm-project/llvm" "-G" "Ninja" "-DLLVM_ENABLE_ASSERTIONS=OFF" "-DLLVM_ENABLE_PLUGINS=OFF" "-DLLVM_TARGETS_TO_BUILD=AArch64;ARM;BPF;Hexagon;MSP430;Mips;NVPTX;PowerPC;RISCV;Sparc;SystemZ;WebAssembly;X86" "-DLLVM_EXPERIMENTAL_TARGETS_TO_BUILD=AVR;M68k" "-DLLVM_INCLUDE_EXAMPLES=OFF" "-DLLVM_INCLUDE_DOCS=OFF" "-DLLVM_INCLUDE_BENCHMARKS=OFF" "-DLLVM_INCLUDE_TESTS=OFF" "-DLLVM_ENABLE_TERMINFO=OFF" "-DLLVM_ENABLE_LIBEDIT=OFF" "-DLLVM_ENABLE_BINDINGS=OFF" "-DLLVM_ENABLE_Z3_SOLVER=OFF" "-DLLVM_PARALLEL_COMPILE_JOBS=8" "-DLLVM_TARGET_ARCH=x86_64" "-DLLVM_DEFAULT_TARGET_TRIPLE=x86_64-pc-windows-msvc" "-DLLVM_INSTALL_UTILS=ON" "-DLLVM_BUILD_INSTRUMENTED=IR" "-DLLVM_PROFILE_DATA_DIR=\\tmp\\tmp-multistage\\llvm-pgo\\prof-%p" "-DLLVM_BUILD_RUNTIME=No" "-DLLVM_ENABLE_ZSTD=OFF" "-DLLVM_ENABLE_ZLIB=OFF" "-DLLVM_USE_CRT_DEBUG=MT" "-DLLVM_USE_CRT_RELEASE=MT" "-DLLVM_USE_CRT_RELWITHDEBINFO=MT" "-DLLVM_ENABLE_LIBXML2=OFF" "-DLLVM_VERSION_SUFFIX=-rust-1.68.0-nightly" "-DCMAKE_INSTALL_MESSAGE=LAZY" "-DCMAKE_C_COMPILER=D:/a/rust/rust/build/bootstrap/debug/sccache-plus-cl.exe" "-DCMAKE_CXX_COMPILER=D:/a/rust/rust/build/bootstrap/debug/sccache-plus-cl.exe" "-DCMAKE_C_FLAGS=-nologo -MT -Brepro --target=x86_64-pc-windows-msvc" "-DCMAKE_CXX_FLAGS=-nologo -MT -Brepro --target=x86_64-pc-windows-msvc" "-DCMAKE_SHARED_LINKER_FLAGS=" "-DCMAKE_MODULE_LINKER_FLAGS=" "-DCMAKE_EXE_LINKER_FLAGS=" "-DCMAKE_INSTALL_PREFIX=D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\llvm" "-DCMAKE_ASM_FLAGS= -nologo -MT -Brepro" "-DCMAKE_ASM_COMPILER=C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.29.30133/bin/HostX64/x64/cl.exe" "-DCMAKE_BUILD_TYPE=Release"
-- The CXX compiler identification is Clang 14.0.5 with MSVC-like command-line
-- The ASM compiler identification is MSVC
-- Found assembler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.29.30133/bin/HostX64/x64/cl.exe
-- Detecting C compiler ABI info
---
    Finished release [optimized] target(s) in 1.98s
Uplifting stage1 library (x86_64-pc-windows-msvc -> x86_64-pc-windows-msvc)
Copying stage2 library from stage1 (x86_64-pc-windows-msvc -> x86_64-pc-windows-msvc / x86_64-pc-windows-msvc)
Build completed successfully in 1:06:47
stage-build INFO: Stage `Build rustc (LLVM PGO)` ended: OK (4007.69s)
stage-build INFO: Stage `Gather profiles (LLVM PGO)` starts
stage-build INFO: Running benchmarks with PGO instrumented LLVM
stage-build DEBUG: Changing working dir from `D:\a\rust\rust` to `D:\a\rust\rust`
stage-build INFO: Executing `RUSTC_BOOTSTRAP=1 D:\a\rust\rust\build\x86_64-pc-windows-msvc\stage2\bin\rustc.exe --edition 2021 --crate-type lib D:\a\rust\rust\library\core\src\lib.rs --out-dir \tmp\tmp-multistage`
stage-build INFO: Executing `RUSTC_BOOTSTRAP=1 D:\a\rust\rust\build\x86_64-pc-windows-msvc\stage2\bin\rustc.exe --edition 2021 --crate-type lib -Copt-level=3 D:\a\rust\rust\library\core\src\lib.rs --out-dir \tmp\tmp-multistage`
stage-build DEBUG: Reverting working dir to `D:\a\rust\rust`
stage-build DEBUG: Changing working dir from `D:\a\rust\rust` to `\tmp\tmp-multistage\rustc-perf`
stage-build INFO: Executing `RUST_LOG=collector=debug RUSTC=D:\tmp\tmp-multistage\rustc-perf\build\x86_64-pc-windows-msvc\stage0\bin\rustc.exe RUSTC_BOOTSTRAP=1 D:\tmp\tmp-multistage\rustc-perf\build\x86_64-pc-windows-msvc\stage0\bin\cargo.exe run -p collector --bin collector -- profile_local eprintln D:\tmp\tmp-multistage\rustc-perf\build\x86_64-pc-windows-msvc\stage2\bin\rustc.exe --id Test --cargo D:\tmp\tmp-multistage\rustc-perf\build\x86_64-pc-windows-msvc\stage0\bin\cargo.exe --profiles Debug,Opt --scenarios Full --include syn-1.0.89,cargo-0.60.0,serde-1.0.136,ripgrep-13.0.0,regex-1.5.5,clap-3.1.6,hyper-0.14.18`
stage-build DEBUG: Reverting working dir to `D:\a\rust\rust`
stage-build INFO: Stage `Gather profiles (LLVM PGO)` ended: FAIL (54.94s)
stage-build ERROR: The multi-stage build has failed
stage-build INFO: Timer results
---------------------------------------------------------
Build rustc (LLVM PGO):                 4007.69s (98.65%)
Gather profiles (LLVM PGO):               54.94s ( 1.35%)
Total duration:                         4062.63s
Traceback (most recent call last):
Traceback (most recent call last):
  File "D:\a\rust\rust\src\ci\stage-build.py", line 644, in <module>
    raise e
  File "D:\a\rust\rust\src\ci\stage-build.py", line 641, in <module>
    execute_build_pipeline(timer, pipeline, build_args)
  File "D:\a\rust\rust\src\ci\stage-build.py", line 581, in execute_build_pipeline
    gather_llvm_profiles(pipeline)
  File "D:\a\rust\rust\src\ci\stage-build.py", line 451, in gather_llvm_profiles
    run_compiler_benchmarks(
  File "D:\a\rust\rust\src\ci\stage-build.py", line 353, in run_compiler_benchmarks
    cmd([
  File "D:\a\rust\rust\src\ci\stage-build.py", line 318, in cmd
    return subprocess.run(args, env=environment, check=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.11.1\x64\Lib\subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.11.1\x64\Lib\subprocess.py", line 1024, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\hostedtoolcache\windows\Python\3.11.1\x64\Lib\subprocess.py", line 1493, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
TypeError: environment can only contain strings
```

</details>

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -439,7 +439,7 @@ jobs:
           - name: dist-x86_64-msvc
             env:
               RUST_CONFIGURE_ARGS: "--build=x86_64-pc-windows-msvc --host=x86_64-pc-windows-msvc --target=x86_64-pc-windows-msvc --enable-full-tools --enable-profiler --set rust.lto=thin"
-              SCRIPT: PGO_HOST=x86_64-pc-windows-msvc src/ci/pgo.sh python x.py dist bootstrap --include-default-paths
+              SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
               DIST_REQUIRE_ALL_TOOLS: 1
             os: windows-latest-xl
           - name: dist-i686-msvc

--- a/src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile
+++ b/src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile
@@ -81,7 +81,7 @@ ENV RUST_CONFIGURE_ARGS \
       --set rust.jemalloc \
       --set rust.use-lld=true \
       --set rust.lto=thin
-ENV SCRIPT ../src/ci/pgo.sh python3 ../x.py dist \
+ENV SCRIPT python3 ../src/ci/stage-build.py python3 ../x.py dist \
     --host $HOSTS --target $HOSTS \
     --include-default-paths \
     build-manifest bootstrap

--- a/src/ci/github-actions/ci.yml
+++ b/src/ci/github-actions/ci.yml
@@ -676,7 +676,7 @@ jobs:
                 --enable-full-tools
                 --enable-profiler
                 --set rust.lto=thin
-              SCRIPT: PGO_HOST=x86_64-pc-windows-msvc src/ci/pgo.sh python x.py dist bootstrap --include-default-paths
+              SCRIPT: PGO_HOST=x86_64-pc-windows-msvc python src/ci/stage-build.py python x.py dist bootstrap --include-default-paths
               DIST_REQUIRE_ALL_TOOLS: 1
             <<: *job-windows-xl
 

--- a/src/ci/stage-build.py
+++ b/src/ci/stage-build.py
@@ -1,0 +1,663 @@
+#!/usr/bin/env python3
+# ignore-tidy-linelength
+
+# Compatible with Python 3.6+
+
+import contextlib
+import getpass
+import glob
+import logging
+import os
+import pprint
+import shutil
+import subprocess
+import sys
+import time
+import traceback
+import urllib.request
+from collections import OrderedDict
+from io import StringIO
+from pathlib import Path
+from typing import Callable, Dict, Iterable, List, Optional, Union
+
+PGO_HOST = os.environ["PGO_HOST"]
+
+LOGGER = logging.getLogger("stage-build")
+
+LLVM_PGO_CRATES = [
+    "syn-1.0.89",
+    "cargo-0.60.0",
+    "serde-1.0.136",
+    "ripgrep-13.0.0",
+    "regex-1.5.5",
+    "clap-3.1.6",
+    "hyper-0.14.18"
+]
+
+RUSTC_PGO_CRATES = [
+    "externs",
+    "ctfe-stress-5",
+    "cargo-0.60.0",
+    "token-stream-stress",
+    "match-stress",
+    "tuple-stress",
+    "diesel-1.4.8",
+    "bitmaps-3.1.0"
+]
+
+LLVM_BOLT_CRATES = LLVM_PGO_CRATES
+
+
+class Pipeline:
+    # Paths
+    def checkout_path(self) -> Path:
+        """
+        The root checkout, where the source is located.
+        """
+        raise NotImplementedError
+
+    def downloaded_llvm_dir(self) -> Path:
+        """
+        Directory where the host LLVM is located.
+        """
+        raise NotImplementedError
+
+    def build_root(self) -> Path:
+        """
+        The main directory where the build occurs.
+        """
+        raise NotImplementedError
+
+    def build_artifacts(self) -> Path:
+        return self.build_root() / "build" / PGO_HOST
+
+    def rustc_stage_0(self) -> Path:
+        return self.build_artifacts() / "stage0" / "bin" / "rustc"
+
+    def cargo_stage_0(self) -> Path:
+        return self.build_artifacts() / "stage0" / "bin" / "cargo"
+
+    def rustc_stage_2(self) -> Path:
+        return self.build_artifacts() / "stage2" / "bin" / "rustc"
+
+    def opt_artifacts(self) -> Path:
+        raise NotImplementedError
+
+    def llvm_profile_dir_root(self) -> Path:
+        return self.opt_artifacts() / "llvm-pgo"
+
+    def llvm_profile_merged_file(self) -> Path:
+        return self.opt_artifacts() / "llvm-pgo.profdata"
+
+    def rustc_perf_dir(self) -> Path:
+        return self.opt_artifacts() / "rustc-perf"
+
+    def build_rustc_perf(self):
+        raise NotImplementedError()
+
+    def rustc_profile_dir_root(self) -> Path:
+        return self.opt_artifacts() / "rustc-pgo"
+
+    def rustc_profile_merged_file(self) -> Path:
+        return self.opt_artifacts() / "rustc-pgo.profdata"
+
+    def rustc_profile_template_path(self) -> Path:
+        """
+        The profile data is written into a single filepath that is being repeatedly merged when each
+        rustc invocation ends. Empirically, this can result in some profiling data being lost. That's
+        why we override the profile path to include the PID. This will produce many more profiling
+        files, but the resulting profile will produce a slightly faster rustc binary.
+        """
+        return self.rustc_profile_dir_root() / "default_%m_%p.profraw"
+
+    def supports_bolt(self) -> bool:
+        raise NotImplementedError
+
+    def llvm_bolt_profile_merged_file(self) -> Path:
+        return self.opt_artifacts() / "bolt.profdata"
+
+
+class LinuxPipeline(Pipeline):
+    def checkout_path(self) -> Path:
+        return Path("/checkout")
+
+    def downloaded_llvm_dir(self) -> Path:
+        return Path("/rustroot")
+
+    def build_root(self) -> Path:
+        return self.checkout_path() / "obj"
+
+    def opt_artifacts(self) -> Path:
+        return Path("/tmp/tmp-multistage/opt-artifacts")
+
+    def build_rustc_perf(self):
+        # /tmp/rustc-perf comes from the Dockerfile
+        shutil.copytree("/tmp/rustc-perf", self.rustc_perf_dir())
+        cmd(["chown", "-R", f"{getpass.getuser()}:", self.rustc_perf_dir()])
+
+        with change_cwd(self.rustc_perf_dir()):
+            cmd([self.cargo_stage_0(), "build", "-p", "collector"], env=dict(
+                RUSTC=str(self.rustc_stage_0()),
+                RUSTC_BOOTSTRAP="1"
+            ))
+
+    def supports_bolt(self) -> bool:
+        return True
+
+
+class WindowsPipeline(Pipeline):
+    def __init__(self):
+        self.checkout_dir = Path(os.getcwd())
+
+    def checkout_path(self) -> Path:
+        return self.checkout_dir
+
+    def downloaded_llvm_dir(self) -> Path:
+        return self.checkout_path() / "citools" / "clang-rust"
+
+    def build_root(self) -> Path:
+        return self.checkout_path()
+
+    def opt_artifacts(self) -> Path:
+        return self.checkout_path() / "opt-artifacts"
+
+    def rustc_stage_0(self) -> Path:
+        return super().rustc_stage_0().with_suffix(".exe")
+
+    def cargo_stage_0(self) -> Path:
+        return super().cargo_stage_0().with_suffix(".exe")
+
+    def rustc_stage_2(self) -> Path:
+        return super().rustc_stage_2().with_suffix(".exe")
+
+    def build_rustc_perf(self):
+        # rustc-perf version from 2022-07-22
+        perf_commit = "3c253134664fdcba862c539d37f0de18557a9a4c"
+        rustc_perf_zip_path = self.opt_artifacts() / "perf.zip"
+
+        def download_rustc_perf():
+            download_file(
+                f"https://github.com/rust-lang/rustc-perf/archive/{perf_commit}.zip",
+                rustc_perf_zip_path
+            )
+            with change_cwd(self.opt_artifacts()):
+                unpack_archive(rustc_perf_zip_path)
+                move_path(Path(f"rustc-perf-{perf_commit}"), self.rustc_perf_dir())
+                delete_file(rustc_perf_zip_path)
+
+        retry_action(download_rustc_perf, "Download rustc-perf")
+
+        with change_cwd(self.rustc_perf_dir()):
+            cmd([self.cargo_stage_0(), "build", "-p", "collector"], env=dict(
+                RUSTC=str(self.rustc_stage_0()),
+                RUSTC_BOOTSTRAP="1"
+            ))
+
+    def rustc_profile_template_path(self) -> Path:
+        """
+        On Windows, we don't have enough space to use separate files for each rustc invocation.
+        Therefore, we use a single file for the generated profiles.
+        """
+        return self.rustc_profile_dir_root() / "default_%m.profraw"
+
+    def supports_bolt(self) -> bool:
+        return False
+
+
+class Timer:
+    def __init__(self):
+        # We want this dictionary to be ordered by insertion.
+        # We use `OrderedDict` for compatibility with older Python versions.
+        self.stages = OrderedDict()
+
+    @contextlib.contextmanager
+    def stage(self, name: str):
+        assert name not in self.stages
+
+        start = time.time()
+        exc = None
+        try:
+            LOGGER.info(f"Stage `{name}` starts")
+            yield
+        except BaseException as exception:
+            exc = exception
+            raise
+        finally:
+            end = time.time()
+            duration = end - start
+            self.stages[name] = duration
+            if exc is None:
+                LOGGER.info(f"Stage `{name}` ended: OK ({duration:.2f}s)")
+            else:
+                LOGGER.info(f"Stage `{name}` ended: FAIL ({duration:.2f}s)")
+
+    def print_stats(self):
+        total_duration = sum(self.stages.values())
+
+        # 57 is the width of the whole table
+        divider = "-" * 57
+
+        with StringIO() as output:
+            print(divider, file=output)
+            for (name, duration) in self.stages.items():
+                pct = (duration / total_duration) * 100
+                name_str = f"{name}:"
+                print(f"{name_str:<34} {duration:>12.2f}s ({pct:>5.2f}%)", file=output)
+
+            total_duration_label = "Total duration:"
+            print(f"{total_duration_label:<34} {total_duration:>12.2f}s", file=output)
+            print(divider, file=output, end="")
+            LOGGER.info(f"Timer results\n{output.getvalue()}")
+
+
+@contextlib.contextmanager
+def change_cwd(dir: Path):
+    """
+    Temporarily change working directory to `dir`.
+    """
+    cwd = os.getcwd()
+    LOGGER.debug(f"Changing working dir from `{cwd}` to `{dir}`")
+    os.chdir(dir)
+    try:
+        yield
+    finally:
+        LOGGER.debug(f"Reverting working dir to `{cwd}`")
+        os.chdir(cwd)
+
+
+def move_path(src: Path, dst: Path):
+    LOGGER.info(f"Moving `{src}` to `{dst}`")
+    shutil.move(src, dst)
+
+
+def delete_file(path: Path):
+    LOGGER.info(f"Deleting file `{path}`")
+    os.unlink(path)
+
+
+def delete_directory(path: Path):
+    LOGGER.info(f"Deleting directory `{path}`")
+    shutil.rmtree(path)
+
+
+def unpack_archive(archive: Path):
+    LOGGER.info(f"Unpacking archive `{archive}`")
+    shutil.unpack_archive(archive)
+
+
+def download_file(src: str, target: Path):
+    LOGGER.info(f"Downloading `{src}` into `{target}`")
+    urllib.request.urlretrieve(src, str(target))
+
+
+def retry_action(action, name: str, max_fails: int = 5):
+    LOGGER.info(f"Attempting to perform action `{name}` with retry")
+    for iteration in range(max_fails):
+        LOGGER.info(f"Attempt {iteration + 1}/{max_fails}")
+        try:
+            action()
+            return
+        except:
+            LOGGER.error(f"Action `{name}` has failed\n{traceback.format_exc()}")
+
+    raise Exception(f"Action `{name}` has failed after {max_fails} attempts")
+
+
+def cmd(
+        args: List[Union[str, Path]],
+        env: Optional[Dict[str, str]] = None,
+        output_path: Optional[Path] = None
+):
+    args = [str(arg) for arg in args]
+
+    environment = os.environ.copy()
+
+    cmd_str = ""
+    if env is not None:
+        environment.update(env)
+        cmd_str += " ".join(f"{k}={v}" for (k, v) in (env or {}).items())
+        cmd_str += " "
+    cmd_str += " ".join(args)
+    if output_path is not None:
+        cmd_str += f" > {output_path}"
+    LOGGER.info(f"Executing `{cmd_str}`")
+
+    if output_path is not None:
+        with open(output_path, "w") as f:
+            return subprocess.run(
+                args,
+                env=environment,
+                check=True,
+                stdout=f
+            )
+    return subprocess.run(args, env=environment, check=True)
+
+
+def run_compiler_benchmarks(
+        pipeline: Pipeline,
+        profiles: List[str],
+        scenarios: List[str],
+        crates: List[str],
+        env: Optional[Dict[str, str]] = None
+):
+    env = env if env is not None else {}
+
+    # Compile libcore, both in opt-level=0 and opt-level=3
+    with change_cwd(pipeline.build_root()):
+        cmd([
+            pipeline.rustc_stage_2(),
+            "--edition", "2021",
+            "--crate-type", "lib",
+            str(pipeline.checkout_path() / "library/core/src/lib.rs"),
+            "--out-dir", pipeline.opt_artifacts()
+        ], env=dict(RUSTC_BOOTSTRAP="1", **env))
+
+        cmd([
+            pipeline.rustc_stage_2(),
+            "--edition", "2021",
+            "--crate-type", "lib",
+            "-Copt-level=3",
+            str(pipeline.checkout_path() / "library/core/src/lib.rs"),
+            "--out-dir", pipeline.opt_artifacts()
+        ], env=dict(RUSTC_BOOTSTRAP="1", **env))
+
+    # Run rustc-perf benchmarks
+    # Benchmark using profile_local with eprintln, which essentially just means
+    # don't actually benchmark -- just make sure we run rustc a bunch of times.
+    with change_cwd(pipeline.rustc_perf_dir()):
+        cmd([
+            pipeline.cargo_stage_0(),
+            "run",
+            "-p", "collector", "--bin", "collector", "--",
+            "profile_local", "eprintln",
+            pipeline.rustc_stage_2(),
+            "--id", "Test",
+            "--cargo", pipeline.cargo_stage_0(),
+            "--profiles", ",".join(profiles),
+            "--scenarios", ",".join(scenarios),
+            "--include", ",".join(crates)
+        ], env=dict(
+            RUST_LOG="collector=debug",
+            RUSTC=str(pipeline.rustc_stage_0()),
+            RUSTC_BOOTSTRAP="1",
+            **env
+        ))
+
+
+# https://stackoverflow.com/a/31631711/1107768
+def format_bytes(size: int) -> str:
+    """Return the given bytes as a human friendly KiB, MiB or GiB string."""
+    KB = 1024
+    MB = KB ** 2  # 1,048,576
+    GB = KB ** 3  # 1,073,741,824
+    TB = KB ** 4  # 1,099,511,627,776
+
+    if size < KB:
+        return f"{size} B"
+    elif KB <= size < MB:
+        return f"{size / KB:.2f} KiB"
+    elif MB <= size < GB:
+        return f"{size / MB:.2f} MiB"
+    elif GB <= size < TB:
+        return f"{size / GB:.2f} GiB"
+    else:
+        return str(size)
+
+
+# https://stackoverflow.com/a/63307131/1107768
+def count_files(path: Path) -> int:
+    return sum(1 for p in path.rglob("*") if p.is_file())
+
+
+def count_files_with_prefix(path: Path) -> int:
+    return sum(1 for p in glob.glob(f"{path}*") if Path(p).is_file())
+
+
+# https://stackoverflow.com/a/55659577/1107768
+def get_path_size(path: Path) -> int:
+    if path.is_dir():
+        return sum(p.stat().st_size for p in path.rglob("*"))
+    return path.stat().st_size
+
+
+def get_path_prefix_size(path: Path) -> int:
+    """
+    Get size of all files beginning with the prefix `path`.
+    Alternative to shell `du -sh <path>*`.
+    """
+    return sum(Path(p).stat().st_size for p in glob.glob(f"{path}*"))
+
+
+def get_files(directory: Path, filter: Optional[Callable[[Path], bool]] = None) -> Iterable[Path]:
+    for file in os.listdir(directory):
+        path = directory / file
+        if filter is None or filter(path):
+            yield path
+
+
+def build_rustc(
+        pipeline: Pipeline,
+        args: List[str],
+        env: Optional[Dict[str, str]] = None
+):
+    arguments = [
+                    sys.executable,
+                    pipeline.checkout_path() / "x.py",
+                    "build",
+                    "--target", PGO_HOST,
+                    "--host", PGO_HOST,
+                    "--stage", "2",
+                    "library/std"
+                ] + args
+    cmd(arguments, env=env)
+
+
+def create_pipeline() -> Pipeline:
+    if sys.platform == "linux":
+        return LinuxPipeline()
+    elif sys.platform in ("cygwin", "win32"):
+        return WindowsPipeline()
+    else:
+        raise Exception(f"Optimized build is not supported for platform {sys.platform}")
+
+
+def gather_llvm_profiles(pipeline: Pipeline):
+    LOGGER.info("Running benchmarks with PGO instrumented LLVM")
+    run_compiler_benchmarks(
+        pipeline,
+        profiles=["Debug", "Opt"],
+        scenarios=["Full"],
+        crates=LLVM_PGO_CRATES
+    )
+
+    profile_path = pipeline.llvm_profile_merged_file()
+    LOGGER.info(f"Merging LLVM PGO profiles to {profile_path}")
+    cmd([
+        pipeline.downloaded_llvm_dir() / "bin" / "llvm-profdata",
+        "merge",
+        "-o", profile_path,
+        pipeline.llvm_profile_dir_root()
+    ])
+
+    LOGGER.info("LLVM PGO statistics")
+    LOGGER.info(f"{profile_path}: {format_bytes(get_path_size(profile_path))}")
+    LOGGER.info(
+        f"{pipeline.llvm_profile_dir_root()}: {format_bytes(get_path_size(pipeline.llvm_profile_dir_root()))}")
+    LOGGER.info(f"Profile file count: {count_files(pipeline.llvm_profile_dir_root())}")
+
+    # We don't need the individual .profraw files now that they have been merged
+    # into a final .profdata
+    delete_directory(pipeline.llvm_profile_dir_root())
+
+
+def gather_rustc_profiles(pipeline: Pipeline):
+    LOGGER.info("Running benchmarks with PGO instrumented rustc")
+
+    # Here we're profiling the `rustc` frontend, so we also include `Check`.
+    # The benchmark set includes various stress tests that put the frontend under pressure.
+    run_compiler_benchmarks(
+        pipeline,
+        profiles=["Check", "Debug", "Opt"],
+        scenarios=["All"],
+        crates=RUSTC_PGO_CRATES,
+        env=dict(
+            LLVM_PROFILE_FILE=str(pipeline.rustc_profile_template_path())
+        )
+    )
+
+    profile_path = pipeline.rustc_profile_merged_file()
+    LOGGER.info(f"Merging Rustc PGO profiles to {profile_path}")
+    cmd([
+        pipeline.build_artifacts() / "llvm" / "bin" / "llvm-profdata",
+        "merge",
+        "-o", profile_path,
+        pipeline.rustc_profile_dir_root()
+    ])
+
+    LOGGER.info("Rustc PGO statistics")
+    LOGGER.info(f"{profile_path}: {format_bytes(get_path_size(profile_path))}")
+    LOGGER.info(
+        f"{pipeline.rustc_profile_dir_root()}: {format_bytes(get_path_size(pipeline.rustc_profile_dir_root()))}")
+    LOGGER.info(f"Profile file count: {count_files(pipeline.rustc_profile_dir_root())}")
+
+    # We don't need the individual .profraw files now that they have been merged
+    # into a final .profdata
+    delete_directory(pipeline.rustc_profile_dir_root())
+
+
+def gather_llvm_bolt_profiles(pipeline: Pipeline):
+    LOGGER.info("Running benchmarks with BOLT instrumented LLVM")
+    run_compiler_benchmarks(
+        pipeline,
+        profiles=["Check", "Debug", "Opt"],
+        scenarios=["Full"],
+        crates=LLVM_BOLT_CRATES
+    )
+
+    merged_profile_path = pipeline.llvm_bolt_profile_merged_file()
+    profile_files_path = Path("/tmp/prof.fdata")
+    LOGGER.info(f"Merging LLVM BOLT profiles to {merged_profile_path}")
+
+    profile_files = sorted(glob.glob(f"{profile_files_path}*"))
+    cmd([
+        "merge-fdata",
+        *profile_files,
+    ], output_path=merged_profile_path)
+
+    LOGGER.info("LLVM BOLT statistics")
+    LOGGER.info(f"{merged_profile_path}: {format_bytes(get_path_size(merged_profile_path))}")
+    LOGGER.info(
+        f"{profile_files_path}: {format_bytes(get_path_prefix_size(profile_files_path))}")
+    LOGGER.info(f"Profile file count: {count_files_with_prefix(profile_files_path)}")
+
+
+def clear_llvm_files(pipeline: Pipeline):
+    """
+    Rustbuild currently doesn't support rebuilding LLVM when PGO options
+    change (or any other llvm-related options); so just clear out the relevant
+    directories ourselves.
+    """
+    LOGGER.info("Clearing LLVM build files")
+    delete_directory(pipeline.build_artifacts() / "llvm")
+    delete_directory(pipeline.build_artifacts() / "lld")
+
+
+def print_binary_sizes(pipeline: Pipeline):
+    bin_dir = pipeline.build_artifacts() / "stage2" / "bin"
+    binaries = get_files(bin_dir)
+
+    lib_dir = pipeline.build_artifacts() / "stage2" / "lib"
+    libraries = get_files(lib_dir, lambda p: p.suffix == ".so")
+
+    paths = sorted(binaries) + sorted(libraries)
+    with StringIO() as output:
+        for path in paths:
+            path_str = f"{path.name}:"
+            print(f"{path_str:<30}{format_bytes(path.stat().st_size):>14}", file=output)
+        LOGGER.info(f"Rustc binary size\n{output.getvalue()}")
+
+
+def execute_build_pipeline(timer: Timer, pipeline: Pipeline, final_build_args: List[str]):
+    # Clear and prepare tmp directory
+    shutil.rmtree(pipeline.opt_artifacts(), ignore_errors=True)
+    os.makedirs(pipeline.opt_artifacts(), exist_ok=True)
+
+    pipeline.build_rustc_perf()
+
+    # Stage 1: Build rustc + PGO instrumented LLVM
+    with timer.stage("Build rustc (LLVM PGO)"):
+        build_rustc(pipeline, args=[
+            "--llvm-profile-generate"
+        ], env=dict(
+            LLVM_PROFILE_DIR=str(pipeline.llvm_profile_dir_root() / "prof-%p")
+        ))
+
+    with timer.stage("Gather profiles (LLVM PGO)"):
+        gather_llvm_profiles(pipeline)
+
+    clear_llvm_files(pipeline)
+    final_build_args += [
+        "--llvm-profile-use",
+        pipeline.llvm_profile_merged_file()
+    ]
+
+    # Stage 2: Build PGO instrumented rustc + LLVM
+    with timer.stage("Build rustc (rustc PGO)"):
+        build_rustc(pipeline, args=[
+            "--rust-profile-generate",
+            pipeline.rustc_profile_dir_root()
+        ])
+
+    with timer.stage("Gather profiles (rustc PGO)"):
+        gather_rustc_profiles(pipeline)
+
+    clear_llvm_files(pipeline)
+    final_build_args += [
+        "--rust-profile-use",
+        pipeline.rustc_profile_merged_file()
+    ]
+
+    # Stage 3: Build rustc + BOLT instrumented LLVM
+    if pipeline.supports_bolt():
+        with timer.stage("Build rustc (LLVM BOLT)"):
+            build_rustc(pipeline, args=[
+                "--llvm-profile-use",
+                pipeline.llvm_profile_merged_file(),
+                "--llvm-bolt-profile-generate",
+            ])
+        with timer.stage("Gather profiles (LLVM BOLT)"):
+            gather_llvm_bolt_profiles(pipeline)
+
+        clear_llvm_files(pipeline)
+        final_build_args += [
+            "--llvm-bolt-profile-use",
+            pipeline.llvm_bolt_profile_merged_file()
+        ]
+
+    # Stage 4: Build PGO optimized rustc + PGO/BOLT optimized LLVM
+    with timer.stage("Final build"):
+        cmd(final_build_args)
+
+
+if __name__ == "__main__":
+    logging.basicConfig(
+        level=logging.DEBUG,
+        format="%(name)s %(levelname)-4s: %(message)s",
+    )
+
+    LOGGER.info(f"Running multi-stage build using Python {sys.version}")
+    LOGGER.info(f"Environment values\n{pprint.pformat(dict(os.environ), indent=2)}")
+
+    build_args = sys.argv[1:]
+
+    timer = Timer()
+    pipeline = create_pipeline()
+    try:
+        execute_build_pipeline(timer, pipeline, build_args)
+    except BaseException as e:
+        LOGGER.error("The multi-stage build has failed")
+        raise e
+    finally:
+        timer.print_stats()
+
+    print_binary_sizes(pipeline)

